### PR TITLE
LIDO-3077 : 2D Array

### PR DESCRIPTION
**Pull request by marhong2**
Updated, should accept 2D arrays in string parsing. @tyleradams @jmorganmartin

### CODE REVIEW

File: grammar/parsing.js [hunk 1]
```diff
@@ -206,12 +206,47 @@ class Parsing extends EmbeddedActionsParser {
...
+                    multiArr.push($.SUBRULE($.subArray))
+                    return false
+                }},
+                {ALT: () => {
+                    singleArr = $.SUBRULE($.constantArray);
+                    $.CONSUME(CloseCurlyParen)
+                    return true;
+                }},
+            ])
+            if(single === true){
```
Comment (tyleradams) [inline grammar/parsing.js:225]: `if(single)`

**Review by tyleradams — CHANGES_REQUESTED** [review]
Please add test cases

File: grammar/parsing.js (hunk 1 above)
Comment (marhong2) [reply grammar/parsing.js:225]: when assessing or statements, chevrotain runs "simulations" on the code to pick the best path (its black boxed). when it does this, the $.OR statement returns a string, rather than either of my Booleans and thus I need to check if it === true. This leads to situations where single is a string, even though the "path" chevrotain took is the false path.

TLDR: chevrotain does a weird thing where it returns a non Boolean to single and thus I need to === true b/c sometimes its a string.

Comment (tyleradams) [reply grammar/parsing.js:228]: Oof that's nasty, thanks for the comment!

File: test/grammar/test.js [hunk 2]
```diff
@@ -66,6 +66,10 @@ describe('Parser allows returning array or range', () => {
         let actual = parser.parse('E5', position, true);
         expect(actual).to.eq(null);
     });
+    it('should work with multiArrays', function() {
```
Comment (tyleradams) [inline test/grammar/test.js:69]: Can we also get some nested formulas. Esp want to make sure 
```
{TRANSPOSE({1,2}),TRANSPOSE({3,4})}
```
works

**Review by tyleradams — APPROVED** [review]

File: grammar/parsing.js (hunk 1 above)
Comment (jmorganmartin) [reply grammar/parsing.js:228]: Very nice code comment. Couple nits on the whitespace (this is consistent throughout the app): 

```suggestion
            if (single === true) {
```

Similarly, spaces after the comments: 
```
// chevrotain occasionally ...
```

File: grammar/parsing.js [hunk 3]
```diff
@@ -206,12 +206,50 @@ class Parsing extends EmbeddedActionsParser {
             return result;
         });
 
+        $.RULE('multiArray', () => {
+            var multiArr = null;
+            let singleArr = null;
```
Comment (jmorganmartin) [inline grammar/parsing.js:211]: Very minor stuff, but it appears that `multiArr` can be a `const` if you initially declare it as `const multiArr = []`. 

This makes L215 unnecessary and probably also the `this.utils.toArray()`s (even if it wasn't a no-op, which it currently is if you look at `utils.js`)

 Objects and arrays (which are Objects) can almost be declared as `const` unless they are fully reassigned (so the `mutliArr.push()` is OK on a `const`, but `singleArr = $.SUBRULE` needs a `let`, as you have done.)

File: grammar/parsing.js [hunk 4]
```diff
@@ -206,12 +206,50 @@ class Parsing extends EmbeddedActionsParser {
...
+                $.CONSUME(Comma);
+                multiArr.push($.SUBRULE($.subArray));
+            })
+            $.CONSUME(CloseCurlyParen)
+            return $.ACTION(() => this.utils.toArray(multiArr));
+        })
+  
+        $.RULE('subArray', () => {
+            $.CONSUME(OpenCurlyParen)
+            let subArray = $.SUBRULE2($.constantArray);
```
Comment (jmorganmartin) [inline grammar/parsing.js:241]:
```suggestion
            const subArray = $.SUBRULE2($.constantArray);
```

File: grammar/parsing.js [hunk 5]
```diff
@@ -206,12 +206,50 @@ class Parsing extends EmbeddedActionsParser {
...
+            if(subArray.length > 1){
```
Comment (jmorganmartin) [inline grammar/parsing.js:242]: Whitespace: 

```suggestion
            if (subArray.length > 1) {
```

**Review by jmorganmartin — CHANGES_REQUESTED** [review]
This looks great! 

Requesting some minor changes for code consistency (none of them should affect functionality). 

Please also address Tyler's request for a test to add a test for the specific case mentioned in the ticket. This is especially important since manual testing will require manually linking the packages.
